compiler/rustc_smir/src/stable_mir/ty.rs

```diff
@@ -266,6 +275,84 @@ pub struct Allocation {
     pub mutability: Mutability,
 }
 
+impl Allocation {
+    /// Creates new empty `Allocation` from given `Align`.
+    fn new_empty_allocation(align: rustc_target::abi::Align) -> Allocation {
+        Allocation {
+            bytes: Vec::new(),
+            provenance: ProvenanceMap { ptrs: Vec::new() },
+            align: align.bytes(),
+            mutability: Mutability::Not,
+        }
+    }
+}
+
+// We need this method instead of a Stable implementation
+// because we need to get `Ty` of the const we are trying to create, to do that
+// we need to have access to `ConstantKind` but we can't access that inside Stable impl.
+pub fn new_allocation<'tcx>(
+    const_kind: &rustc_middle::mir::ConstantKind<'tcx>,
+    const_value: ConstValue<'tcx>,
+    tables: &mut Tables<'tcx>,
+) -> Allocation {
+    match const_value {
+        ConstValue::Scalar(scalar) => {
+            let size = scalar.size();
+            let align = tables
+                .tcx
+                .layout_of(rustc_middle::ty::ParamEnv::reveal_all().and(const_kind.ty()))
+                .unwrap()
+                .align;
+            let mut allocation = rustc_middle::mir::interpret::Allocation::uninit(size, align.abi);
+            allocation
+                .write_scalar(&tables.tcx, alloc_range(rustc_target::abi::Size::ZERO, size), scalar)
+                .unwrap();
+            allocation.stable(tables)
+        }
+        ConstValue::ZeroSized => {
+            let align = tables
+                .tcx
+                .layout_of(rustc_middle::ty::ParamEnv::empty().and(const_kind.ty()))
+                .unwrap()
+                .align;
+            Allocation::new_empty_allocation(align.abi)
+        }
+        ConstValue::Slice { data, start, end } => {
+            let slice_bytes = data.0.get_bytes_unchecked(AllocRange::from(start..end));
+            let slice_allocation = rustc_middle::mir::interpret::Allocation::from_bytes(
+                slice_bytes,
+                data.0.align,
+                data.0.mutability,
+            );
+            slice_allocation.stable(tables)
+        }
+        ConstValue::ByRef { alloc, offset } => {
+            let size = alloc.0.size();
+            let bytes = alloc.0.get_bytes_unchecked(alloc_range(offset, offset + size));
```

Comment:
for a 10 element slice at offset 4 this is now going to take elements `4..14`, which don't exist.

Also, I just remembered that you may want to use the type size, instead of everything from `offset` to the end.